source/extensions/filters/common/ext_authz/check_request_utils.cc

```diff
@@ -44,6 +44,39 @@ NotHeaderKeyMatcher::NotHeaderKeyMatcher(std::vector<Matchers::StringMatcherPtr>
 
 bool NotHeaderKeyMatcher::matches(absl::string_view key) const { return !matcher_.matches(key); }
 
+// Convenience function for either adding a new header or appending value to existing header.
+void addOrAppendHeader(envoy::config::core::v3::HeaderMap& mutable_header_map,
+                       absl::string_view key, absl::string_view value) {
+  auto* headers = mutable_header_map.mutable_headers();
+  for (auto& header : *headers) {
```

Comment:
Just realized that the field is not a map, and this may have some performance implications.
Consider adding a TODO to create the multimap once before actually iterating through header "map" for each key-value pair (i.e., create the multi-map once, and then serialize it to either the old or new proto field).

Reply:
Actually, headers should not be concatenated at all in the new field.

Reply:
In that case, I am wondering we might want to just use map<string, bytes> for the new header_map to maintain the performance?  Otherwise, iterating the repeated "headers" field in HeaderMap proto is a performance hit.  @ggreenway

Reply:
I don't think the performance is a huge concern. The number of headers in a request shouldn't ever be that large.

And we have to handle the case of duplicate headers, so it would need to be a multimap.